cpp/rng/rng.cpp

```diff
@@ -0,0 +1,171 @@
+//and may have high zero content initially
+static uint64_t hash64(uint64_t val) {
+	val ^= val >> 33;
+	val *= 0xff51afd7ed558ccd;
+	val ^= val >> 33;
+	val *= 0xc4ceb9fe1a85ec53;
+	return val ^ (val >> 33);
+}
+
+static void hash_seed_from_blob(const void *dat, uint64_t *state, int64_t bytes) {
+	const char *data = static_cast<const char *>(dat);
+	uint64_t uint64 = 0;
+	memcpy(&uint64, data, bytes <= 8 ? bytes : 8);
+	state[0] = hash64(uint64);
+	state[1] = hash64(uint64 ^ state[0]);
+	for(int64_t i = 1; i <= bytes - 8; i++) {
+		state[0] = hash64(state[0] ^ *reinterpret_cast<const uint64_t *>(data + i));
+		state[1] = hash64(state[0] ^ state[1]);
+	}
+	
+	if(!(state[0] || state[1])) {
+		state[0] = 0x0123456789abcdef;
+	}
+}
+
+void rng::seed(const void *data, uint32_t bytes) {
+	if(bytes == 0) {
+		throw util::Error("0 bytes supplied as seed data");
+	}
+	hash_seed_from_blob(data, this->state, bytes);
+	this->discard(discard_on_seed);
+}
+
+//http://en.wikipedia.org/wiki/Xorshift
+//!actually does the number crunching
+static inline uint64_t do_rng(uint64_t &s0, uint64_t &s1) {
+	s1 ^= s1 << 23;
+	s1 = (s1 ^ s0 ^ (s1 >> 17) ^ (s0 >> 26));
+	return s1 + s0;
+}
+
+uint64_t rng::random() {
+	std::swap(this->state[0], this->state[1]);
+	return do_rng(this->state[0], this->state[1]);
+}
+
+//This provides serious optimizations over the standard generator by
+//eliminating memory writes and data dependencies
+template<class T, class Lambda>
+static inline void act_fill(T *dat, uint64_t *state, size_t len, const Lambda &op){
+
+	uint64_t s0 = state[0];
+	uint64_t s1 = state[1];
+	size_t ev_len = len - (len & 1);
+	
+	for(size_t i = 0; i < ev_len; i += 2) {
+		op(do_rng(s1, s0), dat, i);
+		op(do_rng(s0, s1), dat, i + 1);
+	}
+	if(len & 1) {
+		op(do_rng(s1, s0), dat, len-1);
+		std::swap(s1, s0);
+	}
+	state[0] = s0;
+	state[1] = s1;
+}
+
+void rng::fill(uint64_t *dat, size_t len) {
+	act_fill(dat, this->state, len, [](uint64_t v, uint64_t *d, size_t i) {d[i] = v;});
+}
+
+void rng::discard(size_t len) {
+	act_fill<uint64_t>(nullptr, this->state, len, [](uint64_t, uint64_t *, size_t) {});
+}
+
+void rng::fill_real(double *dat, size_t len) {
+	act_fill(dat, this->state, len, [](double v, double *d, size_t i) {
+			d[i] = v / rng::max();
+		});
+}
+
+std::ostream &rng::to_stream(std::ostream &out) const {
+	if((out << this->state[0] << " ").good() &&
+	   !(out << this->state[1]).fail()) {
+		return out;
+	}
+	throw util::Error("Unable to write rng to output stream");
+}
+
+std::istream &rng::from_stream(std::istream &in) {
+	if((in >> this->state[0]).good() &&
+	   !(in >> this->state[1]).fail()) {
+		return in;
+	}
+	throw util::Error("Unable to serialize rng from input stream");
+}
+
+std::string rng::to_string() const {
+	std::stringstream strstr;
+	return (strstr << *this).str();
+}
+
+void rng::from_string(const std::string &instr) {
+	std::stringstream strstr(instr);
+	strstr >> *this;
+}
+
+using namespace std::chrono;
```

Comment:
hm, you sure that's a good idea? might be better to use explicitly use the `std::chrono` namespace.